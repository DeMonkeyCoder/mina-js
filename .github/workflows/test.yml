name: Run tests
on:
    push:
        branches:
            - main
jobs:
    my-job:
        name: my-job
        runs-on: ubuntu-latest
        steps:
            - uses: actions/checkout@v4
            - uses: oven-sh/setup-bun@v2
            - run: bun install
            - run: bun run build
            - run: bun run test
