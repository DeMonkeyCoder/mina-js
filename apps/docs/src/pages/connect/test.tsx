import { TestZkApp } from "../../components/test-zkapp";

const TestPage = () => {
	if (import.meta.env.SSR) return null;
	return <TestZkApp />;
};

export default TestPage;
