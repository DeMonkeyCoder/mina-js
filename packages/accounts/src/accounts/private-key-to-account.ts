import {
	NullifierSchema,
	SignedFieldsSchema,
	SignedMessageSchema,
	SignedTransactionSchema,
} from "@mina-js/utils";
import MinaSigner from "mina-signer";
import type { PrivateKeyAccount } from "../types";
import { toAccount } from "./to-account";

export type PrivateKeyToAccountParams = {
	privateKey: string;
	network?: "mainnet" | "testnet";
};

/**
 * @description Creates an Account from a private key.
 *
 * @returns A Private Key Account.
 */
export function privateKeyToAccount({
	privateKey,
	network = "mainnet",
}: PrivateKeyToAccountParams): PrivateKeyAccount {
	const client = new MinaSigner({ network });
	const publicKey = client.derivePublicKey(privateKey);
	const account = toAccount({
		publicKey,
		async signMessage({ message }) {
			return SignedMessageSchema.parse(client.signMessage(message, privateKey));
		},
		async signTransaction({ transaction }) {
			return SignedTransactionSchema.parse(
				client.signTransaction(transaction, privateKey),
			);
		},
		async createNullifier({ message }) {
			return NullifierSchema.parse(
				client.createNullifier(
					message.map((el) => BigInt(el)),
					privateKey,
				),
			);
		},
		async signFields({ fields }) {
			return SignedFieldsSchema.parse(
				client.signFields(
					fields.map((el) => BigInt(el)),
					privateKey,
				),
			);
		},
	});

	return {
		...account,
		source: "privateKey",
	} as PrivateKeyAccount;
}
