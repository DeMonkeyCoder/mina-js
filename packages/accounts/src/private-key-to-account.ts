// import MinaSigner from "mina-signer";

// /**
//  * @description Creates an Account from a private key.
//  *
//  * @returns A Private Key Account.
//  */
// export function privateKeyToAccount(privateKey: string): PrivateKeyAccount {
// 	const client = new MinaSigner({ network: "mainnet" });
// 	const publicKey = client.derivePublicKey(privateKey);
// 	const account = toAccount({
// 		publicKey,
// 		async sign({ hash }) {
// 			return sign({ hash, privateKey, to: "hex" });
// 		},
// 		async signMessage({ message }) {
// 			return signMessage({ message, privateKey });
// 		},
// 		async signTransaction(transaction, { serializer } = {}) {
// 			return signTransaction({ privateKey, transaction, serializer });
// 		},
// 		async signTypedData(typedData) {
// 			return signTypedData({ ...typedData, privateKey });
// 		},
// 	});

// 	return {
// 		...account,
// 		publicKey,
// 		source: "privateKey",
// 	} as PrivateKeyAccount;
// }
