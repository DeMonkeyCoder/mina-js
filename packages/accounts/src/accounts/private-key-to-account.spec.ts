import { expect, it } from "bun:test";
import { accounts } from "../test/constants";
import { privateKeyToAccount } from "./private-key-to-account";

it("matches default values", () => {
	const account = privateKeyToAccount({ privateKey: accounts[0].privateKey });
	expect(account).toMatchSnapshot();
});

it("signs a message", async () => {
	const account = privateKeyToAccount({ privateKey: accounts[0].privateKey });
	const message = "hello world";
	const signedMessage = await account.signMessage({ message });
	expect(signedMessage).toMatchSnapshot();
});

it("signs a transaction", async () => {
	const account = privateKeyToAccount({ privateKey: accounts[0].privateKey });
	const transaction = {
		nonce: 1n,
		from: "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
		to: "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
		amount: 3000000000n,
		fee: 100000000n,
	};
	const signedTransaction = await account.signTransaction({ transaction });
	expect(signedTransaction).toMatchSnapshot();
});

it("creates a nullifier", async () => {
	const account = privateKeyToAccount({ privateKey: accounts[0].privateKey });
	const message = [1n, 2n, 3n];
	const nullifier = await account.createNullifier({ message });
	expect(typeof nullifier.private.c).toBe("bigint");
});

it("signs fields", async () => {
	const account = privateKeyToAccount({ privateKey: accounts[0].privateKey });
	const fields = [1n, 2n, 3n];
	const signedFields = await account.signFields({ fields });
	expect(signedFields).toMatchSnapshot();
});
