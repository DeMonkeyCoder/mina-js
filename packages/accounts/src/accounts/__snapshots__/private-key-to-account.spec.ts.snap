// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`matches default values 1`] = `
{
  "createNullifier": [Function: AsyncFunction],
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signFields": [Function: AsyncFunction],
  "signMessage": [Function: AsyncFunction],
  "signTransaction": [Function: AsyncFunction],
  "source": "privateKey",
  "type": "local",
}
`;

exports[`signs a message 1`] = `
{
  "data": "hello world",
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signature": {
    "field": "26035766330205137973322965816888613275878847328888669700853142209707796713851",
    "scalar": "19081530095096912440933327980723294881571166019208938317545493475572505570273",
  },
}
`;

exports[`signs a transaction 1`] = `
{
  "data": {
    "amount": "3000000000",
    "fee": "100000000",
    "from": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
    "memo": "",
    "nonce": "1",
    "to": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
    "validUntil": "4294967295",
  },
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signature": {
    "field": "14481941291402553573872495627690275042434233928114864565201875644822981729719",
    "scalar": "22662322359617803596615419576835284136875294231266919676620712008667696490687",
  },
}
`;

exports[`signs fields 1`] = `
{
  "data": [
    "1",
    "2",
    "3",
  ],
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signature": "7mXCUvhLhFvG9ptrdfNceCrpThkCUyg1ct2z8uwY7eQbKz7UNmhv33TbuDjTznaypJtXRiMJyQWDnf27TH1FSXG7uJHTKAd9",
}
`;

exports[`signs a zkapp command 1`] = `
{
  "data": {
    "feePayer": {
      "fee": "100000000",
      "nonce": "0",
      "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
      "validUntil": null,
    },
    "zkappCommand": {
      "accountUpdates": [],
      "feePayer": {
        "authorization": "7mWy2iLxnZaVmGtCAwVs1EQu8TCdjbSQjQVvUtDFmB9zaiYVet61cN7jDnd5BBqUupQup159H32CV8xAdRc4By5fXLT1T1qi",
        "body": {
          "fee": "100000000",
          "nonce": "0",
          "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
          "validUntil": null,
        },
      },
      "memo": "E4YM2vTHhWEg66xpj52JErHUBU4pZ1yageL4TVDDpTTSsv8mK6YaH",
    },
  },
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signature": "7mWy2iLxnZaVmGtCAwVs1EQu8TCdjbSQjQVvUtDFmB9zaiYVet61cN7jDnd5BBqUupQup159H32CV8xAdRc4By5fXLT1T1qi",
}
`;
